src/engine/Operation.cpp

```diff
@@ -49,6 +50,15 @@ vector<string> Operation::collectWarnings() const {
   return res;
 }
 
+// _____________________________________________________________________________
+void Operation::addWarningOrThrow(std::string warning) {
+  if (RuntimeParameters().get<"throw-on-unbound-variables">()) {
+    throw InvalidSparqlQueryException(std::move(warning));
+  } else {
+    addWarning(std::move(warning));
+  }
+}
+
```

Comment:
As this is "semantically" const, 
you can make the
`warnings_` member mutable (and then the `addWarning  function also const).
The most correct way to do this is then to also make the `const`functions threadsafe, by making the
`warnings_` and `ad_utility::Synchronized<...>` (look it up in the header/ see the usage in tests).

Then you can leave your code in the spatial join as is.